llvm/unittests/CAS/LazyAtomicPointerTest.cpp

```diff
@@ -35,28 +35,25 @@ TEST(LazyAtomicPointer, loadOrGenerate) {
   EXPECT_EQ(Ptr.load(), &Value);
 }
 
-#if (LLVM_ENABLE_THREADS && !defined(_WIN32))
-// FIXME: This test doesn't work on windows.
+#if (LLVM_ENABLE_THREADS)
 TEST(LazyAtomicPointer, BusyState) {
   int Value = 0;
   LazyAtomicPointer<int> Ptr;
   ThreadPool Threads;
 
-  std::mutex BusyStart, BusyEnd;
-  BusyStart.lock();
-  BusyEnd.lock();
+  // Phase to synchronize.
+  // Phase 0 init, Phase 1 busy, Phase 2 finish.
+  std::atomic<int> Phase;
   Threads.async([&]() {
     Ptr.loadOrGenerate([&]() {
-      BusyStart.unlock();
-      while (!BusyEnd.try_lock()) {
```

Comment:
Out of curiosity, even on Linux this was already broken, right?

> If try_lock is called by a thread that already owns the mutex, the behavior is undefined.

There is no guarantee whether the future will start executing its task immediately or not.

Reply:
IMO condition variables would be a better way of fixing this issue to avoid the spinning loops. For example, note how the new implementation won't work if we have a pool with a single thread

Reply:
yes, it was always broken. Condition variable is a good idea but I doubt that works on no thread enabled environment (old implementation also doesn't work on no threads even if the lock works)

Reply:
CVs should work on single threaded environments, since the CV operations yield the mutex when they are not ready

Reply:
Single thread will work but no thread support will not work, which is why it is guarded.